### PR TITLE
Add zoomOffset method

### PR DESCRIPTION
**Pull request by smkuhne**
The zoom offset method centers the zoom-in on the card view-finder and readjusts the image so that all of the outside portions see an even usage.

### CODE REVIEW

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 1]
```diff
@@ -241,6 +241,255 @@ fun Bitmap.zoom(
...
+): Bitmap {
+    require(cardFinder.left > 0 && cardFinder.right < previewSize.width && cardFinder.top > 0 && cardFinder.bottom < previewSize.height) {
+        "Card viewfinder window extends beyond preview boundary"
+    }
+    val newLeft = cardFinder.left * this.width / previewSize.height
+    val newTop = cardFinder.top * this.height / previewSize.height
+    val newRight = cardFinder.right * this.width / previewSize.width
+    val newBottom = cardFinder.bottom * this.height / previewSize.height
+
+    val scaledCardFinderRect = Rect(newLeft, newTop, newRight, newBottom)
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:259]: let's keep these all together using a rect.
```suggestion
    val scaledPreviewSize = previewSize.scaleAndCenterWithin(this.size())
    val previewScale = scaledPreviewSize.width().toFloat() / previewSize.width
    val scaledCardFinderRect = Rect(
        (cardFinder.left * previewScale).roundToInt(),
        (cardFinder.top * previewScale).roundToInt(),
        (cardFinder.right * previewScale).roundToInt(),
        (cardFinder.bottom * previewScale).roundToInt()
    )
```

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 2]
```diff
@@ -241,6 +241,255 @@ fun Bitmap.zoom(
...
+
+    val x = scaledCardFinderRect.centerX()
+    val y = scaledCardFinderRect.centerY()
+
+    require(x + centerSize.width < this.width && x - centerSize.width > 0 && y + centerSize.height < this.height && y - centerSize.height > 0) {
+        "Card viewfinder is to close to the edge of the screen"
+    }
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:266]: Rather than crashing the SDK in this case, let's just do a best-effort. See the above `projectedCardFinderRect`

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 3]
```diff
@@ -241,6 +241,255 @@ fun Bitmap.zoom(
...
+
+    require(toCenterDimension > 0 && toBorderWidth > 0) {
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:268]: this can probable be `require(toCenterDimension >= 0 && toBorderWidth >= 0) {`. This way, we can use this function to crop out the center and border if we so desire.

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 4]
```diff
@@ -241,6 +241,255 @@ fun Bitmap.zoom(
...
+        "Cannot create an image without a border or without a center"
+    }
+
+    val totalHeight = this.height - centerSize.height
+    val totalWidth = this.width - centerSize.width
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:273]:
```suggestion
    val sizeWithoutCenter = Size(
        this.width - centerSize.width,
        this.height - centerSize.width
    )
```

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 5]
```diff
@@ -241,6 +241,255 @@ fun Bitmap.zoom(
...
+    val c3: Pair<Int, Int>
+    val c4: Pair<Int, Int>
+    var tr1: Pair<Int, Int>
+    val tr2: Pair<Int, Int>
+    val tr3: Pair<Int, Int>
+    val tr4: Pair<Int, Int>
+    var tc1: Pair<Int, Int>
+    val tc2: Pair<Int, Int>
+    val tc3: Pair<Int, Int>
+    val tc4: Pair<Int, Int>
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:291]: these names are hard to parse. is `r1 = right1`? is `r1 = row1`? Can you expand these out into full names? Also, do these represent sizes, points, etc? Let's use a class that describes what each of these are.

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 6]
```diff
@@ -241,6 +241,255 @@ fun Bitmap.zoom(
...
+
+    if (x < this.width / 2) {
+        if (y < this.height / 2) {
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:294]: this logic is an absolute bear to read. let's at least split these up into individual functions. Perhaps we can make a single function that takes `x < this.width / 2` and `y < this.height / 2` as parameters?

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 7]
```diff
@@ -241,6 +241,255 @@ fun Bitmap.zoom(
...
+        Rect(c3.first, r2.first, c3.second, r2.second),
+        Rect(c4.first, r2.first, c4.second, r2.second),
+        Rect(c1.first, r3.first, c1.second, r3.second),
+        Rect(c2.first, r3.first, c2.second, r3.second),
+        Rect(c3.first, r3.first, c3.second, r3.second),
+        Rect(c4.first, r3.first, c4.second, r3.second),
+        Rect(c1.first, r4.first, c1.second, r4.second),
+        Rect(c2.first, r4.first, c2.second, r4.second),
+        Rect(c3.first, r4.first, c3.second, r4.second),
+        Rect(c4.first, r4.first, c4.second, r4.second)
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:470]: this is 16 regions, but we're only using a 9x9 grid to perform the resize. This means we're doing 70% more image concatenation operations than we would with a 9x9 grid.

Let's instead break this problem down a little more and keep our 9x9.

**Comment by smkuhne** [comment]
All conversations resolved, updated zoom method to center the viewfinder, changing the image to a 9x16 crop, replacing the remaining background with gray, and then resegmenting regions.

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 8]
```diff
@@ -81,4 +81,92 @@ fun RectF.centerScaled(scaleX: Float, scaleY: Float): RectF {
...
+            this.bottom
+        else
+            this.centerY() + (this.width() * heightRatio / widthRatio / 2)
+    )
+}
+
+/**
+ * Converts a size to rectangle with the top left corner at 0,0
+ */
+fun Size.rect() = Rect(
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:123]: nit: name this `toRect`

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 9]
```diff
@@ -81,4 +81,92 @@ fun RectF.centerScaled(scaleX: Float, scaleY: Float): RectF {
...
+    0,
+    0,
+    this.width,
+    this.height
+)
+
+/**
+ * Return a rect that is the intersection of two other rects
+ */
+fun Rect.intersection(rect: Rect) = Rect(
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:133]: nit: name this `intersectionWith`

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt (hunk 8 above)
Comment (awushensky) [reply scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:123]: We can also remove the implementations in `CardDetectTest` and `SSDOcrTest`

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 10]
```diff
@@ -81,4 +81,92 @@ fun RectF.centerScaled(scaleX: Float, scaleY: Float): RectF {
...
+    if (rect.left > this.left) rect.left else this.left,
+    if (rect.top > this.top) rect.top else this.top,
+    if (rect.right < this.right) rect.right else this.right,
+    if (rect.bottom < this.bottom) rect.bottom else this.bottom
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:137]:
```suggestion
    max(this.left, rect.left),
    max(this.top, rect.top),
    min(this.right, rect.right),
    min(this.bottom, rect.bottom)
```

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt (hunk 9 above)
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:133]: add `@CheckResult`

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt (hunk 8 above)
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:123]: add `@CheckResult`

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 11]
```diff
@@ -81,4 +81,92 @@ fun RectF.centerScaled(scaleX: Float, scaleY: Float): RectF {
...
 
+/**
+ * Scales up a rect
+ */
+fun Rect.scale(scaleX: Float, scaleY: Float) = Rect(
+    this.centerX() - (this.width() * scaleX / 2).toInt(),
+    this.centerY() - (this.height() * scaleY / 2).toInt(),
+    this.centerX() + (this.width() * scaleX / 2).toInt(),
+    this.centerY() + (this.height() * scaleY / 2).toInt()
+)
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:92]: this is essentially the same as `RectF.centerScaled`. Let's rename this to `Rect.centerScaled` to match and move it next to `RectF.centerScaled`

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 12]
```diff
@@ -81,4 +81,92 @@ fun RectF.centerScaled(scaleX: Float, scaleY: Float): RectF {
...
+
+/**
+ * Creates a rectangle of the given aspect ratio, so that the original rect fits directly inside
+ */
+fun Rect.expandToAspectRatio(widthRatio: Int, heightRatio: Int): Rect {
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:97]: There is a similar method named `maxApsectRatioInSize`, which takes a given `Size` and aspect ratio as a `Float` (x/y) and creates a size of aspect ratio that fits within the given size.

Let's rename this similarly `minAspectRatioSurrounding(area: Size, aspectRatio: Float): Size` and then call it like this:

```
minAspectRatioSurrounding(rect.size(), width.toFloat() / height).centeredOn(rect)
```

Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:97]: add `@CheckResult`

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 13]
```diff
@@ -81,4 +81,92 @@ fun RectF.centerScaled(scaleX: Float, scaleY: Float): RectF {
...
+    if (rect.left > this.left) rect.left else this.left,
+    if (rect.top > this.top) rect.top else this.top,
+    if (rect.right < this.right) rect.right else this.right,
+    if (rect.bottom < this.bottom) rect.bottom else this.bottom
+)
+
+/**
+ * Reevalutate the bounds to be relative to another rect
+ */
+fun Rect.relative(rect: Rect) = Rect(
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:143]: it looks like this is just moving a rect by the `left` and `top` of the given rect? Can we instead rename this to `Rect.move(relativeX: Int, relativeY: Int)`?

Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:143]: add `@CheckResult`

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 14]
```diff
@@ -81,4 +81,92 @@ fun RectF.centerScaled(scaleX: Float, scaleY: Float): RectF {
...
+    this.left - rect.left,
+    this.top - rect.top,
+    this.left - rect.left + this.width(),
+    this.top - rect.top + this.height()
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:147]:
```suggestion
    this.right - rect.left,
    this.bottom - rect.top
```

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 15]
```diff
@@ -81,4 +81,92 @@ fun RectF.centerScaled(scaleX: Float, scaleY: Float): RectF {
     )
 }
 
+/**
+ * Scales up a rect
+ */
+fun Rect.scale(scaleX: Float, scaleY: Float) = Rect(
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:87]: add `@CheckResult`

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 16]
```diff
@@ -81,4 +81,92 @@ fun RectF.centerScaled(scaleX: Float, scaleY: Float): RectF {
...
+    /* top */ rect.centerY() - this.height / 2,
+    /* right */ rect.centerX() + this.width / 2,
+    /* bottom */ rect.centerY() + this.height / 2
+)
+
+/**
+ * Takes a relation between a region of interest and a size and projects the region of interest
+ * to that new location
+ */
+fun Size.projectRegionOfInterest(toSize: Size, regionOfInterest: Rect) = Rect(
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:165]: add `@CheckResult`

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 17]
```diff
@@ -86,159 +94,182 @@ fun hasOpenGl31(context: Context): Boolean {
 /**
  * Fragments the image into multiple segments and places them in new segments.
  */
-fun Bitmap.fragment(
-    fromSegments: Array<Rect>,
-    toSegments: Array<Rect>,
-    toSize: Size
+fun Bitmap.rearrangeBySegments(
+    fromSegments: Map<Rect, Rect>,
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:98]: rename: `segmentMap`

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 18]
```diff
@@ -86,159 +94,182 @@ fun hasOpenGl31(context: Context): Boolean {
...
+    futureImageSize: Size
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:99]: let's remove this since we can calculate it based on the `toSegments`

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 19]
```diff
@@ -86,159 +94,182 @@ fun hasOpenGl31(context: Context): Boolean {
...
 ): Bitmap {
-    require(fromSegments.size == toSegments.size) {
-        "Number of source segments does not match number of destination segments"
-    }
-
     val current = this
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:101]: remove this and replace with `this@Bitmap`

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 20]
```diff
@@ -86,159 +94,182 @@ fun hasOpenGl31(context: Context): Boolean {
...
-    val result = Bitmap.createBitmap(toSize.width, toSize.height, this.config)
+    val result = Bitmap.createBitmap(futureImageSize.width, futureImageSize.height, this.config)
     val canvas = Canvas(result)
 
-    (0 until fromSegments.size).map {
-        val image = current.crop(fromSegments[it]).scale(toSegments[it].size())
+    for ((from, to) in fromSegments) {
+        val image = current.crop(from).scale(to.size())
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:106]: nit: rename this to `segment`

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 21]
```diff
@@ -86,159 +94,182 @@ fun hasOpenGl31(context: Context): Boolean {
...
-    }
-
     val current = this
-    val result = Bitmap.createBitmap(toSize.width, toSize.height, this.config)
+    val result = Bitmap.createBitmap(futureImageSize.width, futureImageSize.height, this.config)
     val canvas = Canvas(result)
 
-    (0 until fromSegments.size).map {
-        val image = current.crop(fromSegments[it]).scale(toSegments[it].size())
+    for ((from, to) in fromSegments) {
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:105]: let's use functional programming instead of a for-loop.

```suggestion
    segmentMap.forEach { from, to ->
```

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 22]
```diff
@@ -86,159 +94,182 @@ fun hasOpenGl31(context: Context): Boolean {
...
         )
     }
 
     return result
 }
 
+fun Bitmap.resizeRegion(
+    originalCenterRect: Rect,
+    futureCenterRect: Rect,
+    futureImageSize: Size
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:121]: nit: rename these parameters
```suggestion
    toCenterRect: Rect,
    toImageSize: Size
```

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 23]
```diff
@@ -86,159 +94,182 @@ fun hasOpenGl31(context: Context): Boolean {
...
+    fromSegments: Map<Rect, Rect>,
+    futureImageSize: Size
 ): Bitmap {
-    require(fromSegments.size == toSegments.size) {
-        "Number of source segments does not match number of destination segments"
-    }
-
     val current = this
-    val result = Bitmap.createBitmap(toSize.width, toSize.height, this.config)
+    val result = Bitmap.createBitmap(futureImageSize.width, futureImageSize.height, this.config)
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:102]:
```suggestion
    if (segmentMap.isEmpty) {
        return Bitmap.createBitmap(0, 0, this.config)
    }

    val newImageSize = segmentMap.values.reduce { a, b ->
        Rect(
            min(a.left, b.left),
            min(a.top, b.top),
            max(a.right, b.right),
            max(a.bottom, b.bottom)
        )
    }.size()

    val result = Bitmap.createBitmap(newImageSize.width, newImageSize.height, this.config)
```

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 24]
```diff
@@ -257,6 +288,8 @@ fun Bitmap.crop(crop: Rect): Bitmap {
  */
 fun Bitmap.size(): Size = Size(this.width, this.height)
 
+fun Bitmap.rect(): Rect = Rect(0, 0, this.width, this.height)
+
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:292]: remove this. we can instead use `bitmap.size().toRect()`

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 25]
```diff
@@ -86,159 +94,182 @@ fun hasOpenGl31(context: Context): Boolean {
...
-        Rect(
-            toBorderWidth + toCenterDimension,
-            toBorderWidth + toCenterDimension,
-            (2 * toBorderWidth + toCenterDimension),
-            (2 * toBorderWidth + toCenterDimension)
-        )
+    // Transforms the viewfinder from the preview to the image size
+    val projectedViewFinder = previewSize.projectRegionOfInterest(toSize = this.size(), regionOfInterest = cardFinder)
+    // Creates a square version of the viewfinder
+    val projectedViewFinderSquare = Size(projectedViewFinder.width(), projectedViewFinder.width())
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:239]: this assumes that the viewFinder is always wider than tall. Let's instead use a `max` function to determine the size.

Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:239]: we don't have any guarantee that the `projectedViewFinderSquare` is within the bounds of the new image. Let's make sure it is.

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 26]
```diff
@@ -86,159 +94,182 @@ fun hasOpenGl31(context: Context): Boolean {
...
+    // Creates a rect of a certain aspect ratio surrounding the view finder
+    val aspectRatioCrop = projectedViewFinderSquare.centerOn(projectedViewFinder).expandToAspectRatio(9, 16)
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:241]: let's make the aspect ratio a constant.

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 27]
```diff
@@ -86,159 +94,182 @@ fun hasOpenGl31(context: Context): Boolean {
...
+    val regionMap = croppedImage.resizeRegion(originalCenterRect, futureCenterRect, futureImageSize)
+    // construct the bitmap from the region map
+    return croppedImage.rearrangeBySegments(regionMap, futureImageSize)
+}
+
+/**
+ * Crops and image using originalImageRect and places it on finalImageRect, which is filled with
+ * gray for the best results
+ */
+fun Bitmap.cropWithFill(finalImageRect: Rect, originalImageRect: Rect): Bitmap {
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:256]: we don't need `originalImageRect` since we have a reference to the bitmap already in `this`

Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:256]: add `@CheckResult`

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 28]
```diff
@@ -86,159 +94,182 @@ fun hasOpenGl31(context: Context): Boolean {
...
+            to.left.toFloat(),
+            to.top.toFloat(),
             null
         )
     }
 
     return result
 }
 
+fun Bitmap.resizeRegion(
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:118]: add `@CheckResult`

Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:118]: relocate this to `Layout.kt` and change the signature to be `Size.resizeRegion`. There's no. need to limit this to a bitmap.

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt (hunk 27 above)
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:256]: rename `finalImageRect` to `cropRegion`

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 29]
```diff
@@ -86,159 +94,182 @@ fun hasOpenGl31(context: Context): Boolean {
 /**
  * Fragments the image into multiple segments and places them in new segments.
  */
-fun Bitmap.fragment(
-    fromSegments: Array<Rect>,
-    toSegments: Array<Rect>,
-    toSize: Size
+fun Bitmap.rearrangeBySegments(
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:97]: add `@CheckResult`

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 30]
```diff
@@ -86,159 +94,182 @@ fun hasOpenGl31(context: Context): Boolean {
...
+        futureImageSize.height
+    )
+)
+
 /**
  * Crops a piece from the center of the source image, resizing that to fit the new center dimension,
  * and squeezes the remainder of the image into a border. The returned image is guaranteed to be
  * square.
  */
 fun Bitmap.zoom(
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:229]: add `@CheckResult`

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt (hunk 18 above)
Comment (smkuhne) [reply scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:99]: The reason this is in there is because I have to create the bitmap + canvas at the beginning of the method. Though We could technically iterate through all the segments and find the size that way, that means we would have to iterate through all the segments twice, once to extrapolate the size to make the canvas, and then once more to actually draw onto the canvas. I personally feel that passing in a size will have a negligible performance impact in comparison to extrapolating the size every time.

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt (hunk 19 above)
Comment (smkuhne) [reply scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:101]: @Bitmap is not required, only use this

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 31]
```diff
@@ -36,8 +38,45 @@ fun maxAspectRatioInSize(area: Size, aspectRatio: Float): Size {
 }
 
 /**
+<<<<<<< HEAD
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:41]: looks like some of the rebasing slipped in. remove this.

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 32]
```diff
@@ -36,8 +38,45 @@ fun maxAspectRatioInSize(area: Size, aspectRatio: Float): Size {
...
  * Calculate the maximum [Size] that fits within the [containingSize] and maintains the same aspect ratio as the subject
  * of this method. This is often used to project a preview image onto a full camera image.
+=======
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:44]: remove this

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 33]
```diff
@@ -1,9 +1,11 @@
 package com.getbouncer.scan.framework.util
 
+import android.graphics.Bitmap
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:3]: shouldn't need this in layout.

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 34]
```diff
@@ -36,8 +38,45 @@ fun maxAspectRatioInSize(area: Size, aspectRatio: Float): Size {
...
+}
+
+/**
+ * Calculate the position of the [Size] within the [containingSize]. This makes a few
+ * assumptions:
+ * 1. the [Size] and the [containingSize] are centered relative to each other.
+ * 2. the [Size] and the [containingSize] have the same orientation
+ * 3. the [containingSize] and the [Size] share either a horizontal or vertical field of view
+ * 4. the non-shared field of view must be smaller on the [Size] than the [containingSize]
+>>>>>>> Implement changes
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:79]: remove this

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 35]
```diff
@@ -102,4 +141,161 @@ fun RectF.centerScaled(scaleX: Float, scaleY: Float) = RectF(
...
+@CheckResult
+fun Rect.intersectionWith(rect: Rect) = Rect(
+    max(this.left, rect.left),
+    max(this.top, rect.top),
+    min(this.right, rect.right),
+    min(this.bottom, rect.bottom)
+)
+
+/**
+ * Reevalutate the bounds to be relative to another rect
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:169]: This comment is out of date

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 36]
```diff
@@ -102,4 +141,161 @@ fun RectF.centerScaled(scaleX: Float, scaleY: Float) = RectF(
...
+    regionOfInterest.top * toSize.height / this.height,
+    regionOfInterest.right * toSize.width / this.width,
+    regionOfInterest.bottom * toSize.height / this.height
+)
+
+/**
+ * Resizes a region of the image and places it somewhere else
+ */
+@CheckResult
+fun Bitmap.resizeRegion(
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:195]: make this `Size.resizeRegion`

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 37]
```diff
@@ -125,127 +144,58 @@ fun Bitmap.fragment(
...
+    val projectViewFinderGreatestDimension = max(projectedViewFinder.width(), projectedViewFinder.height())
+    val projectedViewFinderSquare = Size(projectViewFinderGreatestDimension, projectViewFinderGreatestDimension)
+
+    require(this.size().toRect().contains(projectedViewFinderSquare.centerOn(projectedViewFinder))) {
+        "Viewfinder too close to edge of the screen"
+    }
+
+    val aspectRatio = 9.0f / 16.0f
+    // Creates a rect of a certain aspect ratio surrounding the view finder
+    val aspectRatioCrop = minAspectRatioSurroundingSize(projectedViewFinderSquare, aspectRatio).centerOn(projectedViewFinder)
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:165]: this should probably happen outside of zoom. it's a separate step that should occur before we zoom in.

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 38]
```diff
@@ -125,127 +144,58 @@ fun Bitmap.fragment(
  */
 @CheckResult
 fun Bitmap.zoom(
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:146]: this method is starting to get very specific to our screen detect use case. perhaps we should move the logic in to the ScreenDetect class.

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 39]
```diff
@@ -102,4 +141,161 @@ fun RectF.centerScaled(scaleX: Float, scaleY: Float) = RectF(
 )
 
 @CheckResult
+fun Rect.centerScaled(scaleX: Float, scaleY: Float) = Rect(
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:144]: please add a test for this

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 40]
```diff
@@ -102,4 +141,161 @@ fun RectF.centerScaled(scaleX: Float, scaleY: Float) = RectF(
...
+ * Converts a size to rectangle with the top left corner at 0,0
+ */
+@CheckResult
+fun Size.toRect() = Rect(0, 0, this.width, this.height)
+
+/**
+ * Return a rect that is the intersection of two other rects
+ */
+@CheckResult
+fun Rect.intersectionWith(rect: Rect) = Rect(
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:161]: please add a test for this

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 41]
```diff
@@ -102,4 +141,161 @@ fun RectF.centerScaled(scaleX: Float, scaleY: Float) = RectF(
...
+    max(this.top, rect.top),
+    min(this.right, rect.right),
+    min(this.bottom, rect.bottom)
+)
+
+/**
+ * Reevalutate the bounds to be relative to another rect
+ */
+@CheckResult
+fun Rect.move(relativeX: Int, relativeY: Int) = Rect(
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:172]: please add a test for this

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 42]
```diff
@@ -102,4 +141,161 @@ fun RectF.centerScaled(scaleX: Float, scaleY: Float) = RectF(
...
+    this.right - relativeX,
+    this.bottom - relativeY
+)
+
+/**
+ * Takes a relation between a region of interest and a size and projects the region of interest
+ * to that new location
+ */
+@CheckResult
+fun Size.projectRegionOfInterest(toSize: Size, regionOfInterest: Rect) = Rect(
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:184]: please add a test for this

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt (hunk 36 above)
Comment (awushensky) [reply scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:195]: please add a test for this method

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt [hunk 43]
```diff
@@ -125,127 +144,58 @@ fun Bitmap.fragment(
...
+    // construct the bitmap from the region map
+    return croppedImage.rearrangeBySegments(regionMap)
+}
+
+/**
+ * Crops and image using originalImageRect and places it on finalImageRect, which is filled with
+ * gray for the best results
+ */
+@CheckResult
+fun Bitmap.cropWithFill(cropRegion: Rect): Bitmap {
```
Comment (awushensky) [inline scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:181]: please add a unit test

File: scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt [hunk 44]
```diff
@@ -36,8 +38,45 @@ fun maxAspectRatioInSize(area: Size, aspectRatio: Float): Size {
...
+ * ```
+ *  ____
+ * |____|
+ * |    |
+ * |____|
+ * |____|
+ * ```
+ */
+@CheckResult
+fun minAspectRatioSurroundingSize(area: Size, aspectRatio: Float): Size {
```
Comment (awushensky) [inline scan-framework/src/main/java/com/getbouncer/scan/framework/util/Layout.kt:59]: add a test

File: scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt (hunk 24 above)
Comment (awushensky) [reply scan-payment/src/main/java/com/getbouncer/scan/payment/Image.kt:200]: Can this be removed?
